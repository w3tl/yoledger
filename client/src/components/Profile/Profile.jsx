import React from 'react';
import PropTypes from 'prop-types';
import { TextInput, Button } from '../Elements';

class Profile extends React.Component {
  static propTypes = {
    changePassword: PropTypes.func.isRequired,
  }

  state = {
    oldPassword: '',
    newPassword: '',
    repeatNewPassword: '',
  }

  handleChange = name => ({ target }) => {
    this.setState({
      [name]: target.value,
    });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { changePassword } = this.props;
    const { oldPassword, newPassword } = this.state;
    changePassword({ oldPassword, newPassword });
  }

  render() {
    const { oldPassword, newPassword, repeatNewPassword } = this.state;
    const isNewPassword = oldPassword !== newPassword;
    const isRepeatCorrectly = newPassword.length && newPassword === repeatNewPassword;
    const isValid = isNewPassword && isRepeatCorrectly;
    return (
      <div>
        <h4>Change password</h4>
        <form onSubmit={this.handleSubmit}>
          <TextInput name="oldPassword" label="Old password" value={oldPassword} required type="password" onChange={this.handleChange('oldPassword')} />
          <TextInput
            required
            warning={!isNewPassword}
            message="New password should be unique"
            name="newPassword"
            label="New password"
            value={newPassword}
            type="password"
            onChange={this.handleChange('newPassword')}
          />
          <TextInput
            required
            warning={!isRepeatCorrectly}
            message="Repeat password correctly"
            name="repeatNewPassword"
            label="Repeat new password"
            value={repeatNewPassword}
            type="password"
            onChange={this.handleChange('repeatNewPassword')}
          />
          <Button type="submit" disabled={!isValid}>Change</Button>
        </form>
      </div>
    );
  }
}

export default Profile;
