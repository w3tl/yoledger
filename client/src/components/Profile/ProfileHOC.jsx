/* eslint-disable react/destructuring-assignment,react/prop-types */
import React from 'react';
import { Mutation } from 'react-apollo';
import { CHANGE_PASSWORD_MUTATION } from './queries';

const withChangePasswordMutation = Wrapped => props => (
  <Mutation
    mutation={CHANGE_PASSWORD_MUTATION}
    onCompleted={(data) => {
      localStorage.setItem('token', data.changePassword.token);
    }}
  >
    {(changePassword, { loading, error }) => (
      <Wrapped
        changePassword={changePassword}
        loading={loading}
        error={error}
        {...props}
      />)}
  </Mutation>
);

export default Component => withChangePasswordMutation(Component);
