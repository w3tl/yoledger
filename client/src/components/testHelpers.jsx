import React from 'react';
import { shallow, mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

export const withProvider = Component => props => (
  <MockedProvider {...props}>
    <Component />
  </MockedProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return fixedDate;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
