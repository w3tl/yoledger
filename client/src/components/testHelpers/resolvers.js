import dateResolver from './date';
import assets from '../Asset/mockData';
import { expenses, incomes } from '../Categories/mockData';
import { accounts, budgets as rawBudgets } from '../Budget/mockData';
import transactions from '../Transaction/mockData';

const accountsArray = assets.concat(expenses).concat(incomes);
const budgets = rawBudgets.map(({ date, ...fields }) => ({ date: new Date(date), ...fields }));

export default {
  Date: dateResolver,
  Query: {
    account(root, { id }) {
      if (id === '666') return { name: 'new account' };
      return accountsArray.find(a => a.id === id);
    },
    accounts(root, { type }) {
      return accountsArray.filter(a => a.type === type);
    },
    budgets() {
      return { accounts };
    },
    budget(root, { account, dateStart, dateEnd }) {
      return budgets // eslint-disable-next-line max-len
        .filter(b => (
          b.account.name === account
          && b.date.getTime() >= dateStart.getTime()
          && b.date.getTime() <= new Date(dateEnd).getTime()));
    },
    transactions() {
      return transactions.map(({ date, ...other }) => ({ date: new Date(date), ...other }));
    },
    transaction(root, { id }) {
      if (id === '666') return { name: 'new account' };
      const { date, ...other } = transactions.find(a => a.id === id);
      return { date: new Date(date), ...other };
    },
  },
  Mutation: {
    upsertBudget(root, { input: { account, date, amount } }) {
      const existingBudget = budgets
        .find(b => (
          b.account.name === account.name && b.date.getTime() === new Date(date).getTime()
        ));
      const acc = accounts.find(a => a.name === account) || {
        id: '687',
        name: account,
      };
      return {
        success: true,
        budget: {
          id: '99',
          ...existingBudget,
          account: acc,
          date,
          amount,
        },
      };
    },
    addAccount(root, { input }) {
      return {
        account: { id: '666', ...input },
      };
    },
    addTransaction(root, { input }) {
      return {
        transaction: {
          id: '666',
          ...input,
          source: accountsArray.find(a => a.name === input.source),
          destination: accountsArray.find(a => a.name === input.destination),
        },
      };
    },
    deleteTransaction(root, { id }) {
      return {
        success: true,
        id,
      };
    },
    updateTransaction(root, { id, input }) {
      return {
        transaction: {
          ...transactions.find(t => t.id === id),
          id: '888',
          ...input,
          source: accountsArray.find(a => a.name === input.source),
          destination: accountsArray.find(a => a.name === input.destination),
        },
      };
    },
    signin(root, { login }) {
      return {
        token: `{\\"userId\\":\\"${login}\\"};secret;{\\"expiresIn\\":10}`,
      };
    },
    signout() {
      return {
        success: true,
      };
    },
    changePassword() {
      return {
        token: `{\\"userId\\":\\"${'admin'}\\"};secret;{\\"expiresIn\\":10}`,
      };
    },
  },
};
