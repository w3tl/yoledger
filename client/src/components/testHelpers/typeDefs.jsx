export default `
scalar Date
type Query {
  account(id: ID!): Account
  accounts(type: AccountType!): [Account]

  budgets(dateStart: Date!, count: Int = 5): Budget
  budget(date: Date!): [Allocation]

  transactions(dateStart: Date!, dateEnd: Date, page: Int = 0, itemsPerPage: Int = 20): [Transaction]
  transaction(id: ID!): Transaction
}
type Mutation {
  addAccount(input: AddAccountInput!): AddAccountPayload
  upsertBudget(input: UpsertBudgetInput!): UpsertBudgetPayload
  addTransaction(input: AddTransactionInput!): AddTransactionPayload
  deleteTransaction(id: ID!): DeleteTransactionPayload
  updateTransaction(id: ID!, input: UpdateTransactionInput!): UpdateTransactionPayload
}

type Account {
  id: ID!
  name: String!
  balance: Float
  type: AccountType!
}

type Budget {
  periods: [Date]
  accounts: [Account]
}

type Allocation {
  id: String!
  account: Account!
  amount: Float
  balance: Float
}

enum AccountType {
  ASSET
  EXPENSE
  INCOME
  VENDOR
}

input AddAccountInput {
  name: String!
  balance: Float
  type: AccountType!
}

type AddAccountPayload {
  account: Account
}

input UpsertBudgetInput {
  account: String!
  date: Date!
  amount: Float!
}

type UpsertBudgetPayload {
  success: Boolean
  allocation: Allocation
}

type Transaction {
  id: ID!
  amount: Float!
  source: Account!
  destination: Account!
  date: Date!
}
input AddTransactionInput {
  source: String!
  destination: String!
  amount: Float!
  date: Date
}
type AddTransactionPayload {
  transaction: Transaction
}
type DeleteTransactionPayload {
  success: Boolean!
  id: ID
}
input UpdateTransactionInput {
  amount: Float
  source: String
  destination: String
  date: Date
}
type UpdateTransactionPayload {
  transaction: Transaction
}
`;
