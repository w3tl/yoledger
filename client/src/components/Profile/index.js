import Profile from './Profile';
import withData from './ProfileHOC';

export default withData(Profile);
