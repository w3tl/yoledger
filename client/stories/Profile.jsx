import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import Profile from '../src/components/Profile/Profile';

const ComponentWithActions = props => (
  <Profile changePassword={action('changePassword')} {...props} />
);

storiesOf('Profile', module)
  .add('default', () => <ComponentWithActions />)
  .add('with same new password', () => <ComponentWithActions initialState={{ oldPassword: '1', newPassword: '1' }} />)
  .add('with wrong repeated password', () => <ComponentWithActions initialState={{ oldPassword: '1', newPassword: '2', repeatNewPassword: '3' }} />)
  .add('loading', () => <ComponentWithActions loading />)
  .add('with errors', () => <ComponentWithActions error={{ graphQLErrors: ['No user found!'] }} />);
