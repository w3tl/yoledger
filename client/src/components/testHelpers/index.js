import React from 'react';
import { shallow, mount } from 'enzyme';
import { ApolloProvider } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { makeExecutableSchema } from 'graphql-tools';
import typeDefs from './typeDefs';
import resolvers from './resolvers';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

const executableSchema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

const cache = new InMemoryCache();
const link = new SchemaLink({ schema: executableSchema });
export const client = new ApolloClient({
  link,
  cache,
});

export const withProvider = Component => props => (
  <ApolloProvider client={client}>
    <Component client={client} />
  </ApolloProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  const date = new Date(fixedDate);
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return date;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
