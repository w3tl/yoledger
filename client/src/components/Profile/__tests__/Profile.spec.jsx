import React from 'react';
import { mount } from 'enzyme';
import Profile from '../Profile';
import { testRenderWithoutError } from '../../testHelpers';

const Component = props => <Profile changePassword={() => {}} {...props} />;

describe('Profile', () => {
  testRenderWithoutError(<Component />);

  describe('Change button', () => {
    it('should be disabled when the old or new password empty', () => {
      const output = mount(<Component />);
      expect(output.find('Button[type="submit"]').prop('disabled')).toBeTruthy();
    });
    it('should be disabled when the old password are equals to new', () => {
      const output = mount(<Component />);
      output.find('TextInput[name="oldPassword"]').simulate('change', { target: { value: 'password' } });
      expect(output.find('Button[type="submit"]').prop('disabled')).toBeTruthy();
      output.find('TextInput[name="newPassword"]').simulate('change', { target: { value: 'password' } });
      output.find('TextInput[name="repeatNewPassword"]').simulate('change', { target: { value: 'password' } });
      expect(output.find('Button[type="submit"]').prop('disabled')).toBeTruthy();
    });
    it('should be disabled when the new password does not match the repeat', () => {
      const output = mount(<Component />);
      output.find('TextInput[name="oldPassword"]').simulate('change', { target: { value: 'password' } });
      output.find('TextInput[name="newPassword"]').simulate('change', { target: { value: 'password1' } });
      output.find('TextInput[name="repeatNewPassword"]').simulate('change', { target: { value: 'password2' } });
      expect(output.find('Button[type="submit"]').prop('disabled')).toBeTruthy();
    });
    it('should be enabled when all fine', () => {
      const output = mount(<Component />);
      output.find('TextInput[name="oldPassword"]').simulate('change', { target: { value: 'password' } });
      output.find('TextInput[name="newPassword"]').simulate('change', { target: { value: 'password1' } });
      output.find('TextInput[name="repeatNewPassword"]').simulate('change', { target: { value: 'password1' } });
      expect(output.find('Button[type="submit"]').prop('disabled')).toBeFalsy();
    });
  });
  //
  // it('should disable submit button when', () => {
  //   const signin = jest.fn();
  //   const output = mount(
  //     <Login signin={signin} />,
  //   );
  //   output.find('TextInput[name="username"]').simulate('change', { target: { value: 'admin' } });
  //   output.find('TextInput[name="password"]').simulate('change', { target: { value: 'password' } });
  //   output.find('form').simulate('submit');
  //   expect(signin).toHaveBeenCalledWith({
  //     username: 'admin',
  //     password: 'password',
  //   });
  // });
});
