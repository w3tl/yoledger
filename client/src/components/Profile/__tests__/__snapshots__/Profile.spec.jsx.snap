// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile renders without error 1`] = `
<Profile
  changePassword={[Function]}
/>
`;
